### PR TITLE
SONARJS-902 X-proc baby step

### PR DESCRIPTION
**Pull request by vilchik-elena**

### CODE REVIEW

File: javascript-frontend/src/main/java/org/sonar/javascript/se/SymbolicExecution.java [hunk 1]
```diff
@@ -350,12 +382,18 @@ private void afterBlockElement(ProgramState currentState, Tree element) {
...
   private void pushAllSuccessors(CfgBlock block, ProgramState currentState) {
     for (CfgBlock successor : block.successors()) {
-      pushSuccessor(successor, currentState);
+      pushSuccessor(successor, block, currentState);
     }
   }
 
-  private void pushSuccessor(CfgBlock successor, @Nullable ProgramState currentState) {
+  private void pushSuccessor(CfgBlock successor, CfgBlock block, @Nullable ProgramState currentState) {
     if (currentState != null) {
```
Comment (inverno) [inline javascript-frontend/src/main/java/org/sonar/javascript/se/SymbolicExecution.java]: Extract and move this logic at the end of block execution.

File: javascript-frontend/src/test/resources/se/cross_procedure.js [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+      if (condition) {
+        return 5;
+      }
+      // if "condition" is false, function returns "undefined" implicitly
+   }
+
+   barReturn = bar();
+   foo(barReturn); // PS barReturn=UNDEFINED_OR_TRUTHY_NUMBER
+
+
```
Comment (inverno) [inline javascript-frontend/src/test/resources/se/cross_procedure.js]: Add a test for no-expression return.

File: javascript-frontend/src/main/java/org/sonar/javascript/se/SymbolicExecution.java [hunk 3]
```diff
@@ -350,11 +403,11 @@ private void afterBlockElement(ProgramState currentState, Tree element) {
 
   private void pushAllSuccessors(CfgBlock block, ProgramState currentState) {
     for (CfgBlock successor : block.successors()) {
-      pushSuccessor(successor, currentState);
+      pushSuccessor(successor, block, currentState);
     }
   }
 
-  private void pushSuccessor(CfgBlock successor, @Nullable ProgramState currentState) {
+  private void pushSuccessor(CfgBlock successor, CfgBlock block, @Nullable ProgramState currentState) {
```
Comment (SonarTech) [inline javascript-frontend/src/main/java/org/sonar/javascript/se/SymbolicExecution.java]: MAJOR Remove this unused method parameter "block". [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS1172)

**Comment by SonarTech** [comment]
SonarQube analysis reported 1 issue
* ![MAJOR][MAJOR] 1 major

Watch the comments in this conversation to review them.

[MAJOR]: https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png 'Severity: MAJOR'

File: javascript-frontend/src/test/resources/se/cross_procedure.js [hunk 4]
```diff
@@ -49,5 +49,16 @@ function main(array, i) {
    barReturn = bar();
    foo(barReturn); // PS barReturn=UNDEFINED_OR_TRUTHY_NUMBER
 
+   bar = function() {
+      if (condition) {
+        return 5;
+      }
+
+      return; // implicit return of "undefined"
```
Comment (inverno) [inline javascript-frontend/src/test/resources/se/cross_procedure.js]: I would not call this "implicit return" as you already use the word "implicit" in the method that adds `UNDEFINED` in the return constraint when no return is there at all. I would change the comment to read something like "expression-less return of undefined".

**Pull request merged**